File: common/src/main/java/it/futurecraft/sorrentino/http/plugins/VerifySignaturePlugin.java
package it.futurecraft.sorrentino.http.plugins;

import io.javalin.config.JavalinState;
import io.javalin.http.BadRequestResponse;
import io.javalin.http.ForbiddenResponse;
import io.javalin.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.List;
import java.util.function.Consumer;

public class VerifySignaturePlugin extends Plugin<VerifySignaturePlugin.Config> {
    public static class Config {
        public @NotNull String secret;
        public @NotNull List<String> ignorList;
    }

    public VerifySignaturePlugin(@NotNull Consumer<Config> userConfig) {
        super(userConfig, null);
    }

    @Override
    public void onInitialize(@NotNull JavalinState state) {
        state.routes.before(ctx -> {
            if (pluginConfig.ignorList.contains(ctx.path())) return;

            String signature = ctx.header("Twitch-Eventsub-Message-Signature");
            String messageId = ctx.header("Twitch-Eventsub-Message-Id");
            String timestamp = ctx.header("Twitch-Eventsub-Message-Timestamp");

            if (signature == null || messageId == null || timestamp == null) {
                throw new BadRequestResponse("Missing required headers.");
            }

            String body = ctx.body();
            String message = messageId + timestamp + body;

            String expected = hmac(message);

            if (!expected.equals(signature)) {
                throw new ForbiddenResponse("Invalid signature.");
            }
        });
    }

    private String hmac(@NotNull String message) throws NoSuchAlgorithmException, InvalidKeyException {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec key = new SecretKeySpec(pluginConfig.secret.getBytes(), "HmacSHA256");

        mac.init(key);
        byte[] result = mac.doFinal(message.getBytes());

        return HexFormat.of().formatHex(result);
    }
}

File: paper/src/main/kotlin/it/futurecraft/sorrentino/http/plugins/VerifySignaturePlugin.kt
package it.futurecraft.sorrentino.http.plugins

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.experimental.and

internal fun hmac(message: String, secret: String): String {
    val mac = Mac.getInstance("HmacSHA256")
    val key = SecretKeySpec(secret.toByteArray(), "HmacSHA256")

    mac.init(key)
    val result = mac.doFinal(message.toByteArray())

    return result.joinToString("") { "%02x".format(it and 0xff.toByte()) }
}

data class VerifySignaturePluginConfig(
    var secret: String = "",
    var ignore: List<String> = emptyList(),
)

val VerifySignaturePlugin = createApplicationPlugin(
    name = "VerifyRequestPlugin",
    createConfiguration = ::VerifySignaturePluginConfig
) {
    val secret = pluginConfig.secret

    pluginConfig.apply {
        onCall { call ->
            val uri = call.request.uri

            if (ignore.any { uri.startsWith(it) }) {
                return@onCall
            }

            val signature = call.request.headers["Twitch-Eventsub-Message-Signature"]
            val messageId = call.request.headers["Twitch-Eventsub-Message-Id"]
            val timestamp = call.request.headers["Twitch-Eventsub-Message-Timestamp"]

            if (signature == null || timestamp == null || messageId == null) {
                call.respondText("Malformed request headers.", status = HttpStatusCode.BadRequest)
                return@onCall
            }

            val body = call.receiveText()
            val message = "$messageId$timestamp$body"

            val generated = "sha256=${hmac(message, secret)}"

            if (generated != signature) {
                call.respondText("Invalid signature.", status = HttpStatusCode.Forbidden)
                return@onCall
            }
        }
    }
}